### DISCUSSION TITLE
检测系统内是否存在某个命令行程序

### DISCUSSION THREAD

**Discussion by jingkaimori** [xmake-io/xmake#3156] [Ideas]
### 你在什么场景下需要该功能？

现有的`os.run`函数会通过命令行调用系统内的其他程序，现在检测不到系统路径上有没有指定的程序，因此直接运行命令可能会出错，需要容错。

在Windows平台下，qt提供了windeploy这一命令行程序，这个程序会扫描编译好的binary文件，然后把该文件依赖的dll复制到合适的目录，以避免找不到dll的错误。

### 描述可能的解决方案

在os命名空间下添加新函数，对应的命令行程序满足以下条件时，该函数会返回`true`，否则返回`false`
* 命令行程序存在
* 命令行程序可执行
* （可选）可以用`--version`选项测试对应的程序
``` lua
os.commandexist('ls') -- return true
os.commandexist('C:/Qt/5.15.2/mingw/windeploy') -- return true when given qt environment is installed
```

### 描述你认为的候选方案

修改qt相关的规则，编译binary后调用Windows下的`windeploy`或macos下的`macdeploy`，部署对应的动态链接库文件

**Comment by waruqi** [comment]
```lua
import("lib.detect.find_tool")
local gcc = find_tool("gcc")
if gcc then
    os.execv(gcc.program, {"--version"})
end
```

or 

```lua
local ok = try {function () os.run("gcc --help") end}
if ok then
end
```
